File: IAIso-v5.0/core/iaiso-java/iaiso-consent/src/main/java/io/iaiso/consent/Scopes.java
package io.iaiso.consent;

import java.util.List;

/**
 * Scope grammar and matching, per {@code spec/consent/README.md §4–§5}.
 * <pre>
 * scope   ::= segment ("." segment)*
 * segment ::= [a-z0-9_-]+
 * </pre>
 *
 * A token granting {@code G} satisfies a request for {@code R} iff:
 * <ul>
 *   <li>{@code G == R} (exact match), or
 *   <li>{@code R} starts with {@code G + "."} (prefix at segment boundary).
 * </ul>
 */
public final class Scopes {
    private Scopes() {}

    public static boolean granted(List<String> granted, String requested) {
        if (requested == null || requested.isEmpty()) {
            throw new IllegalArgumentException("requested scope must be non-empty");
        }
        for (String g : granted) {
            if (g.equals(requested)) {
                return true;
            }
            if (requested.startsWith(g + ".")) {
                return true;
            }
        }
        return false;
    }
}

File: IAIso-v5.0/core/iaiso-csharp/src/Iaiso.Consent/Scopes.cs
using System;
using System.Collections.Generic;

namespace Iaiso.Consent;

/// <summary>
/// Scope grammar and matching, per <c>spec/consent/README.md §4–§5</c>.
/// </summary>
/// <remarks>
/// <code>
/// scope   ::= segment ("." segment)*
/// segment ::= [a-z0-9_-]+
/// </code>
/// A token granting <c>G</c> satisfies a request for <c>R</c> iff:
/// <list type="bullet">
///   <item><c>G == R</c> (exact match), or</item>
///   <item><c>R</c> starts with <c>G + "."</c> (prefix at segment boundary).</item>
/// </list>
/// </remarks>
public static class Scopes
{
    public static bool Granted(IReadOnlyList<string> granted, string requested)
    {
        if (string.IsNullOrEmpty(requested))
        {
            throw new ArgumentException("requested scope must be non-empty", nameof(requested));
        }
        foreach (var g in granted)
        {
            if (g == requested) return true;
            if (requested.StartsWith(g + ".", StringComparison.Ordinal)) return true;
        }
        return false;
    }
}
